# 프로젝트 (2) load_wine : 와인을 분류해 봅시다

- 두번째 실습 : 제공된 데이터를 활용하여 와인의 특징을 잡고 종류를 분류해 보자.


- Scikit-Learn 라이브러리의 datasets를 모듈 안의 load_wine 매서드 사용해서 프로젝트를 진행

---
### (1) 필요한 모듈 import 하기
---

In [2]:
from sklearn.datasets import load_wine  # 실습파일 불러오기
from sklearn.model_selection import train_test_split  # model_selection 모듈 안의 Train / Test Set을 나누기 위한 함수 불러오기
from sklearn.metrics import classification_report  # metrics 모듈 안의 학습 모델을 평가하기 위한 함수 불러오기
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

import pandas as pd  # pandas 모듈 불러오기
import numpy as np   # numpy 모듈 불러오기
import matplotlib.pyplot as plt

---
### (2) 데이터 준비
---

In [3]:
wine = load_wine()  # load_wine 매서드를 사용해서 wine 데이터 불러옴
type(wine)  # wine의 타입을 구하면 파이썬 딕셔너리, 번치(bunch) 객체로 표현됨

sklearn.utils.Bunch

- `wine.keys()` 매서드를 사용해서 `wine` 객체 안의 정보 확인해보자

In [5]:
wine.keys()  # digits의 속성 확인하기

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

* `wine`의 데이터 정보를 파악하기 위해 pandas 모듈의 DataFrame을 활용해보자

In [9]:
wine_df = pd.DataFrame(data=wine.data, columns=wine.feature_names)
wine_df['label'] = wine.target
wine_df

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  label  
0                            3.92   1065.0      0  
1                            3.40   1050.0      0  
2                            3.17   1185.0      0  
3                            3.45   1480.0      0  
4                            2.93    735.0      0  
..                            ...      ...    ...  
173                          1.74    740.0      2  
174                          1.56    750.0      2  
175                          1.56    835.0      2  
176                          1.62    840.0      2  
177                          1.60    560.0      2  

[178 rows x 14 columns]

In [11]:
wine_df['label'].value_counts()  # wine의 target 값 중 unique한 값들의 중복되는 갯수는 다음과 같다.

1    71
0    59
2    48
Name: label, dtype: int64

* `wine`의 feature data의 크기는 178x13 이며 총 178개의 샘플 데이터가 있으며 13개의 특성을 갖는다.


* label은 ['class_0' 'class_1' 'class_2']로 나누어진다.


* feature는 <br>['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']<br>로 나누어진다.

---
### (3) 데이터 이해하기
---

**1) Feature Data 지정하기**


In [12]:
wine_data = wine.data  # Feature Data 지정
print(wine_data.shape)   # Feature Data의 모양 출력
print(type(wine_data))   # Feature Data의 객체 타입 출력

(178, 13)
<class 'numpy.ndarray'>


**2) Label Data 지정하기**

In [13]:
wine_label = wine.target  # Label Data 지정
print(wine_label.shape)     # Label Data의 모양 출력
print(type(wine_label))     # Label Data의 객체 타입 출력

(178,)
<class 'numpy.ndarray'>


**3) Target Names 출력해보기**

In [14]:
wine_target_names = wine.target_names
print(wine_target_names)

['class_0' 'class_1' 'class_2']


**4) 데이터 Describe 해보기**

In [15]:
print(wine.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

---
### (4) train, test 데이터 분리
---

- 모델 학습과 테스트용 문제지와 정답지를 준비해보자.
    - `train_test_split` 함수를 사용하여 Data를 학습 : 테스트의 비율을 8:2로 골고루 섞이도록 나눠보자


In [33]:
X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, 
                                                    random_state=1234)
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(142, 13) (142,)
(36, 13) (36,)


- 아래 셀의 결과를 확인하면 나누어진 Test Data가 전체 Data에서 골고루 섞여져 있는 것으로 확인된다.

In [34]:
unique, counts = np.unique(y_test, return_counts=True)
dict(zip(unique, counts))


{0: 10, 1: 17, 2: 9}

---
### (5) 다양한 모델로 학습시켜보기
---

- 1) **Decision Tree** 사용해 보기<br>
- 2) **Random Forest** 사용해 보기<br>
- 3) **SVM** 사용해 보기<br>
- 4) **SGD Classifier** 사용해 보기<br>
- 5) **Logistic Regression** 사용해 보기<br>

In [44]:
# 학습모델 불러오기 

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import LogisticRegression

In [58]:
# 학습모델을 각각의 변수에 저장

decision_tree = DecisionTreeClassifier(random_state=1234)
random_forest = RandomForestClassifier(random_state=1234)
svm_model = svm.SVC()
sgd_model = SGDClassifier()
logistic_model = LogisticRegression()

In [46]:
# 각각의 모델을 Train Data set로 학습

decision_tree.fit(X_train, y_train)
random_forest.fit(X_train, y_train)
svm_model.fit(X_train, y_train)
sgd_model.fit(X_train, y_train)
logistic_model.fit(X_train, y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [47]:
# Test Data를 통해 학습모델의 결과값 예측

prediction_decision_tree = decision_tree.predict(X_test)
prediction_random_forest = random_forest.predict(X_test)
prediction_svm_model = svm_model.predict(X_test)
prediction_sgd_model = sgd_model.predict(X_test)
prediction_logistic_model = logistic_model.predict(X_test)

---
### (6) 모델을 평가해 보기
---

- **1) sklearn.metrics accuracy 지표를 사용한 경우**

In [57]:
print(accuracy_score(y_test, prediction_decision_tree))
print(accuracy_score(y_test, prediction_random_forest))
print(accuracy_score(y_test, prediction_svm_model))
print(accuracy_score(y_test, prediction_sgd_model))
print(accuracy_score(y_test, prediction_logistic_model))
# print(accuracy_score(y_test, [1]*len(prediction_logistic_model)))

0.8888888888888888
0.9444444444444444
0.6944444444444444
0.6388888888888888
0.8888888888888888


- **2) sklearn.metrics classification_report 지표를 사용한 경우**

In [39]:
print(classification_report(y_test, prediction_decision_tree), end='------------------------------------------------------------\n\n')
print(classification_report(y_test, prediction_random_forest), end='------------------------------------------------------------\n\n')
print(classification_report(y_test, prediction_svm_model), end='------------------------------------------------------------\n\n')
print(classification_report(y_test, prediction_sgd_model), end='------------------------------------------------------------\n\n')
print(classification_report(y_test, prediction_logistic_model), end='------------------------------------------------------------\n\n')

              precision    recall  f1-score   support

           0       1.00      0.80      0.89        10
           1       0.88      0.88      0.88        17
           2       0.82      1.00      0.90         9

    accuracy                           0.89        36
   macro avg       0.90      0.89      0.89        36
weighted avg       0.90      0.89      0.89        36
------------------------------------------------------------

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      0.88      0.94        17
           2       0.82      1.00      0.90         9

    accuracy                           0.94        36
   macro avg       0.94      0.96      0.95        36
weighted avg       0.95      0.94      0.95        36
------------------------------------------------------------

              precision    recall  f1-score   support

           0       1.00      0.70      0.82        10
       

- **3) sklearn.metrics confusion_matrix 지표를 사용한 경우**

In [43]:
print(confusion_matrix(y_test, prediction_decision_tree), end='\n\n')
print(confusion_matrix(y_test, prediction_random_forest), end='\n\n')
print(confusion_matrix(y_test, prediction_svm_model), end='\n\n')
print(confusion_matrix(y_test, prediction_sgd_model), end='\n\n')
print(confusion_matrix(y_test, prediction_logistic_model))

[[ 8  2  0]
 [ 0 15  2]
 [ 0  0  9]]

[[10  0  0]
 [ 0 15  2]
 [ 0  0  9]]

[[ 7  2  1]
 [ 0 17  0]
 [ 0  8  1]]

[[ 7  0  3]
 [ 0  3 14]
 [ 0  0  9]]

[[ 9  1  0]
 [ 1 14  2]
 [ 0  0  9]]


---
### (7) 결론
---

- `프로젝트(2) load_wine : 와인을 분류해봅시다`에서는 178개의 데이터 샘플을 가지고 총 13개의 feature(특징값)를 사용하여 와인의 클래스 [0,1,2]를 맞추기위한 **지도학습 - 분류 문제**라 할 수 있다. 


- 총 178개의 Data Sample 중 필자는 train_test_split 함수를 이용하여 Data Sample을 8:2 비율로 균등하게 나누었다. 아래 그림과 같이 Data Sample의 데이터들의 치우침 정도를 확인하기 위해 pandas DataFrame의 매소드 value_counts()를 사용하였다.<br> **※ 데이터가 불균형(imbalance) 하지 않음을 확인하였다**


![image](https://user-images.githubusercontent.com/103712369/165583296-a845e3b8-7386-4dda-9349-bde76a5a12ad.png)

- 본 노드를 통해 배운 Scikit-Learn에서 제공하는 평가지표 함수는 accuracy / classification_report / confusion_matrix가 있었는데 confusion_matrix의 성능지표 중 가장 많이 사용되는 Precision / Recall을 사용하기에는 데이터가 고루 분포되어 있기 때문에 **Accuracy 성능지표를 사용하여 학습모델을 평가하는것이 바람직하다고 생각된다.**


- 모델은 **Ensemble 기법 중 하나인 RandomForestClassifier**를 사용했을 때 약 94%로 가장 좋은 성능을 보였다. 
